### PR TITLE
fix(training): fix compile error when using activation checkpointing

### PR DESCRIPTION
**Pull request by cathalobrien**
## Description
Fixes an error when using torch compiled code inside a checkpointed region.

the error was reported [here](https://github.com/ecmwf/anemoi-core/issues/1330) with more background [here](https://github.com/pytorch/pytorch/issues/166926). The fix comes from [here](https://github.com/pytorch/torchtitan/pull/1991) and is only available for torch 2.10 and above. I have added a warning if you use compilation and torch < 2.10 which lists some workarounds. but the proper solution is upgrading torch.

It also fixes another cause of recompilation errors: `torch._inductor.config.shape_padding` can pad tensors based on micro-benchmarks. if you recompile inside a checkpointed reason this can lead to shape mismatches. I disable this feature when using gradient checkpointing. I also warn users about other settings e.g. "mode='max-autotune'" which could lead to errors when used with gradient checkpointing. 

Lastly, I set a default number of CPU threads. Otherwise this was being reset multiple times by PytorchLightning during runtime. Each time it was set, it would trigger a needless recompilation, which used up one of the 'model.recompile_limit' recompilations

I got Claude to generate a reproducer, using the compiled conditional layer norm in anemoi (see [here](https://github.com/ecmwf/anemoi-core/blob/3cb72df1f1767b8588b518b0c6e9af98f3e4893c/reproduce_compile_checkpoint_lru.py)) the error fires when the cache is enabled, and doesn't fire when the cache is disabled.

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/Contributor-License-Agreement.md)

### CODE REVIEW

File: training/src/anemoi/training/utils/compile.py [hunk 1]
```diff
@@ -102,6 +103,20 @@ def prepare_compilation(
...
+
+    # disable LRU cache, this is a fix for https://github.com/pytorch/pytorch/issues/166926
+    # The runtime impact of this should be marginal
+    if version.parse(torch.__version__) >= version.parse("2.10.0"):
+        torch._C._dynamo.eval_frame._set_lru_cache(False)
+    else:
+        LOGGER.warning(
+            "Could not disable torch compile LRU cache because torch version is < 2.10.0. "
+            "This may result in runtime errors when using torch.compile() alongside activation checkpointing. "
+            "If you encounter errors, consider upgrading to torch >= 2.10.0, disabling torch.compile() (model.compile=[]) "
```
Comment (ssmmnn11) [inline training/src/anemoi/training/utils/compile.py]:
```suggestion
            "If you encounter errors, consider either upgrading to torch >= 2.10.0, or disabling torch.compile() (model.compile=[]) "
```

**Review by ssmmnn11 — DISMISSED** [review]
Excellent. Minor suggestion only.

**Review by VeraChristina — DISMISSED** [review]
Thanks for fixing this! LGTM

**Review by VeraChristina — DISMISSED** [review]

**Review by VeraChristina — APPROVED** [review]
Disabling torch compile in the edm tests, which were the ones still failing with this error, seems like a good temporary solution (as that it also the solution suggested to the user in the logs)
